src/WebJobs.Script.WebHost/Middleware/HostWarmupMiddleware.cs

```diff
@@ -33,21 +37,31 @@ public HostWarmupMiddleware(RequestDelegate next, IScriptWebHostEnvironment webH
             _assemblyLocalPath = Path.GetDirectoryName(new Uri(typeof(HostWarmupMiddleware).Assembly.Location).LocalPath);
         }
 
-        public async Task Invoke(HttpContext httpContext)
+        public Task Invoke(HttpContext httpContext)
         {
             if (IsWarmUpRequest(httpContext.Request, _webHostEnvironment.InStandbyMode, _environment))
             {
-                PreJitPrepare(WarmUpConstants.JitTraceFileName);
-                if (_environment.IsLinuxConsumption())
-                {
-                    PreJitPrepare(WarmUpConstants.LinuxJitTraceFileName);
-                }
+                return WarmupInvoke(httpContext);
+            }
 
-                ReadRuntimeAssemblyFiles();
+            return _next.Invoke(httpContext);
+        }
 
-                await WarmUp(httpContext.Request);
+        /// <summary>
+        /// This is so we only pay the async overhead while in the warmup path, but not for primary runtime.
+        /// </summary>
+        public async Task WarmupInvoke(HttpContext httpContext)
+        {
+            PreJitPrepare(WarmUpConstants.JitTraceFileName);
+            if (_environment.IsLinuxConsumption())
+            {
+                PreJitPrepare(WarmUpConstants.LinuxJitTraceFileName);
```

Comment:
oh we call PreJitPrepare 2 times for Linux, with this update we are basically skipping the 2nd call because by that time _jitTraceHasRun is true.

Reply:
oh yes...what do we _intend_ to do there? Are we intending to actually create 2 of the same file, or should we just be swapping the filename?

Reply:
For Linux we want to call the method 2 times, one with  WarmUpConstants.JitTraceFileName which is generic for both Windows and Linux and 2nd time with WarmUpConstants.LinuxJitTraceFileName which is specific to Linux.

maybe move the _jitTraceHasRun check in WarmupInvoke and set it when both are called?

Reply:
Done!